transaction/DirectedGraph.hpp

```diff
@@ -0,0 +1,173 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TRANSACTION_DIRECTED_GRAPH_HPP_
+#define QUICKSTEP_TRANSACTION_DIRECTED_GRAPH_HPP_
+
+#include <algorithm>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <stack>
+#include <vector>
+
+#include "transaction/Transaction.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+namespace transaction {
+
+/** \addtogroup Transaction
+ *  @{
+ */
+
+/**
+ * @brief Class for representing a directed graph.
+ *        Vertices are transaction ids, edges are
+ *        wait-for relations.
+ */
+class DirectedGraph {
+ public:
+  using NodeId = std::uint64_t;
+
+  /**
+   * @brief Default constructor
+   */
+  DirectedGraph() {}
+
+  /**
+   * @brief Adds node with given transaction id.
+   * @warning It does not check whether transaction id
+   *          is already in the graph.
+   * @warning Pointer ownership will pass to the graph,
+   *          threfore it should not be deleted.
+   *
+   * @param data Pointer to the transaction id that
+   *        will be contained in the node.
+   *
+   * @return Id of the newly created node.
+   */
+  inline NodeId addNode(TransactionId *data) {
+    nodes_.emplace_back(data);
+    return nodes_.size() - 1;
+  }
+
+  /**
+   * @brief Adds an edge between nodes.
+   * @warning Does not check arguments are legit.
+   *          It may cause out of range errors.
+   *
+   * @param fromNode The node that edge is orginated.
+   * @param toNode The node that edge is ended.
+   */
+  inline void addEdge(NodeId from_node, NodeId to_node) {
+    nodes_[from_node].addOutgoingEdge(to_node);
+  }
+
+  /**
+   * @brief Check whether there is a directed edge.
+   * @warning Does not check argument are legit.
+   *          It may cause out of range errors.
+   *
+   * @param fromNode Id of the node that edge is originated from.
+   * @param toNode Id of the node that edge is ended.
+   *
+   * @return True if there is an edge, false otherwise.
+   */
+  inline bool hasEdge(NodeId from_node, NodeId to_node) const {
+    return nodes_[from_node].hasOutgoingEdge(to_node);
+  }
+
+  /**
+   * @brief Get data (transaction id) contained in the node.
+   * @warning Does not check index validity.
```

Comment:
You can add a DCHECK to ensure that the node ID belongs to the graph. DCHECK only runs in debug builds and doesn't cause an overhead in the release build. It is immensely helpful in spotting errors.